### ISSUE TITLE
Busybox Bionic Support

### ISSUE THREAD

**Issue by koroshiya** [deajan/osync#40]
ie. Cyanogen OS (and other Android OS's) support.

osync already syncs to Android, but when it comes to cleaning up on Busybox Bionic shells, osync starts to fail.
I've encountered 2 issues preventing the cleanup from working on BBB: ctime and xargs.

ctime can be worked around by replacing the flag with mtime (which is probably preferable anyway?), and xargs can be replaced with a "while read" loop.

eg.
$FIND_CMD "$replica_deletion_path/" -type f -ctime +$change_time -print0 | xargs -0 -I {} echo "Will delete file {}" > "$RUN_DIR/osync.$FUNCNAME.$SCRIPT_PID"
Logger "Command output:\n$(cat $RUN_DIR/osync.$FUNCNAME.$SCRIPT_PID)" "NOTICE"
$FIND_CMD "$replica_deletion_path/" -type d -empty -ctime +$change_time -print0 | xargs -0 -I {} echo "Will delete directory {}" > "$RUN_DIR/osync.$FUNCNAME.$SCRIPT_PID"
Logger "Command output:\n$(cat $RUN_DIR/osync.$FUNCNAME.$SCRIPT_PID)" "NOTICE"

could be replaced with:
$FIND_CMD "$replica_deletion_path/" -type f -mtime +$change_time -print0 | while read foo; do Logger "Command output:\nWill delete file $foo" "NOTICE"; done
$FIND_CMD "$replica_deletion_path/" -type d -empty -mtime +$change_time -print0 | while read foo; do Logger "Command output:\nWill delete directory $foo" "NOTICE"; done

As shown in the example, this would also clean up the "ugly xarg hacks" used for logging command output.

**Comment by koroshiya** [comment]
It seems there was one more issue: the -P flag in df.
This didn't stop the script from executing, but it did prevent it from retrieving a valid disk space.

I've submitted a pull request which deals with all of the above. It should maintain complete compatibility with non-Busybox systems, but any and all feedback is welcome.

**Comment by deajan** [comment]
I must investigate if those changes are compatible with Linux, FreeBSD, MacOS X and MSYS before being able to merge anything.
Also, replacing ctime with mtime might create problems depending on the way rsync handles backup copies (ie does mtime change when a file is moved ?).

I'll schedule this for in one or two weeks probably.

**Comment by koroshiya** [comment]
fwiw, testing was conducted between Ubuntu (14.04) and Android (5.1.1).
I'll see if I have a device running another version of busybox. I've currently tried it with 1.22.1 and 1.21.1.

Hm, it might cause issues with backup copies then. mtime doesn't change when a file is moved (or permissions are updated). If nothing else the flag could change between ctime and mtime, as the patch currently does with df -P. That would make backups work as they currently do for non-busybox shells, but wouldn't help busybox.

**Comment by koroshiya** [comment]
I should be able to work around the mtime/ctime issue by using stat %Z. It's not as elegant as mtime/ctime, but should be almost completely consistent across shells, and provides the same time as ctime.

*Busybox's stat %Z rounds to the nearest second, so at worst a backup would be deleted one day earlier than intended if the file was modified a fraction of a second before midnight.

**Comment by koroshiya** [comment]
Made this change in: https://github.com/koroshiya/osync/commit/7003c58004751fa339ed0e3c6286b8a6b8a09610
Will test tomorrow. In theory it should have the same functionality as ctime, but with Busybox support.

**Comment by deajan** [comment]
Just looked at your code, it's very ingenious but it won't work for remote deletion because the function DaysHavePassed only exists within the local shell, not on the remote side via ssh.

How about a one liner that would do the same trick ? Can you check if this works in busybox ?
```
find /home/git/obackup/ -type f -exec stat -c "%Z %n" "{}" \; | while read epoch filename; do if [ $epoch -gt $(($(date +"%s") - ( 30 * 24 * 60 * 60))) ]; then echo "This file is older than 30 days: $filename"; fi ; done
```
30 * 24 * 60 * 60 = days * hours * minutes * seconds, in this case, 30 days.

**Comment by koroshiya** [comment]
Nice, that does the trick. But you have the comparison around the wrong way; epoch should be less than the date 30 days ago, since lower epoch = further in the past.

**Comment by deajan** [comment]
I have digged a bit more into your code, and there is more to correct to be mergeable.
You use "df" instead of "df -P" for busybox that you detect locally, but this would not work in case of a BSD to busybox sync for example.
That's where GetLocalOS, GetRemoteOS, InitLocalOSSettings and InitRemoteOSSettings come in handy. Depending on the OS, different commands can be set there.

**Comment by koroshiya** [comment]
Will do.
Remote/Local OS is straightforward enough, though detection here will be a bit more tricky.
It seems to be an Android thing rather than Busybox, and Android's uname reports as being Linux.

**Comment by deajan** [comment]
No problem. Was an interesting discussion though. I'll close PR#40.
Let me know how you can detect BBB, se if we have to add other commands to uname detection.

**Comment by deajan** [comment]
Btw, I have commited a lot of small changes I made to osync that shares code with another project so your next PR would be based on something new.

**Comment by koroshiya** [comment]
Will do.
I've started from scratch (using the latest osync.sh) and taken a different approach.
I'm now checking for Busybox Bionic, then using different commands for it, so it doesn't mess with other platforms.
The only change I've made affecting other platforms was to move the SSH command from an escaped string into a text file.

I think it should be good to go: https://github.com/koroshiya/osync/commits/master
However, I'm going to do a bit more testing before considering sending a pull request.

**Comment by deajan** [comment]
I've added some comments to your commits.
What was the problem with the find oneliner I gave you ?
I honestly wouldn't like to have a bunch of txt files that come with my code.

**Comment by koroshiya** [comment]
The txt files could be replaced by actual in-line heredocs. I just moved them into txt files because lengthy heredocs look messy in nested commands like this.

The main problem with the one-liner was mentioned earlier: https://github.com/deajan/osync/issues/40#issuecomment-155730197
Although it works on regular Busybox, it doesn't work on BBB.

**Comment by deajan** [comment]
I understood that the oneliner didn't work, but I did not understand what was the issue.
Is it because of the -c flag with stat ? (sorry, bad english understanding some times)

Heck I'm interested into some testing, but never worked with busybox until now.
Can you tell me what's a quick and dirty way to get a test environment for BBB ?

**Comment by deajan** [comment]
Oh... a last thing.
The osync.sh in the git root is generated by dev/merge.sh that combines dev/ofunctions.sh into dev/n_osync.sh

All your work should go into n_osync.sh (and the ConvertUnits function should go into ofunctions.sh  as it is generic enough to be in the function library).
Please work with those files so I won't have to "migrate" your PR later.

**Comment by koroshiya** [comment]
Will do. I'll merge it all into those sh files once everything is satisfactory.


IIRC, the problem was the formats accepted by the -c flag. I think the flag is there, but with extremely limited arguments accepted.

If you have a spare android device, you can flash a nightly CyanogenMod 12.1 ROM on it. That's my current test environment for BBB.

You should be able to install BBB on a Linux system (or Android device) by following the instructions on git: https://github.com/CyanogenMod/android_external_busybox/blob/cm-12.0/INSTALL
However, I have not attempted this myself, so I can't say.

You should be able to get a VM running a CM device.
Genymotion may have one.
If not, you could try using Android studio, as mentioned here: https://stackoverflow.com/questions/8820127/cyanogenmod-on-android-emulator-is-it-possible

**Comment by deajan** [comment]
Okay, I'll see if I can have a BBB running on a linux VM this weekend, so I can play with it.
Btw, what's the purpose of running osync on a phone ?

**Comment by koroshiya** [comment]
Local BBB support is just for the hell of it. It's remote support I'm interested in.
ie. Using osync to sync data to and from a phone from a PC.

My use case is offline data syncing/backup between two computers and a phone. So the phone acts as both a backup device and an intermediary through which the two computers can sync.

**Comment by deajan** [comment]
Anything to merge yet ?

**Comment by koroshiya** [comment]
Not yet. If you're waiting on this for the next version, I'd say to release without it.

There's an error I'd like to fix before merging anything, but I won't have time to look at it until at least next week.
It's an error with deleted files on Android, where files that have already been deleted from disk still show up in the list. (It doesn't break anything, but it does lead to "errors" in the script execution)

**Comment by koroshiya** [comment]
I've sorted that error (problem with the heredocs), moved the heredocs from separate files into the script itself, and instead made the changes to n_osync.sh and ofunctions.sh
I grabbed the current code and manually applied my own changes in a way that should only affect BBB devices (ie. shouldn't change anything for Linux, Mac, Windows, etc.)

However, the merge.sh file seems to reference build files on your own PC, so I can't build osync.sh.
eg. /home/git/common/common_install.sh
For the time being I've committed the changes to my fork.
